src/core/reactor_backend.cc

```diff
@@ -882,6 +884,18 @@ reactor_backend_epoll::wait_and_process(int timeout, const sigset_t* active_sigm
             auto op = evt.events ? EPOLL_CTL_MOD : EPOLL_CTL_DEL;
             ::epoll_ctl(_epollfd.get(), op, pfd->fd.get(), &evt);
         }
+        else if (has_error && pfd->events_requested == 0) {
```

Comment:
I'm having trouble understanding all the details of this method.

`events_to_remove` de-registers (`EPOLL_CTL_MOD`) events that are are triggered (in `evt.events`) but only the _next_ time through the loop?

That is, if `EPOLLIN` is requested and gets triggered, it isn't going to be removed from the fd list until the next time through this method (since it's not removed from `pfd->events_requested` until inside `complete_epoll_event` which happens after the `auto events_to_remove = ...` calculation). Is that right and do you understand why that is?

That seems to be the reason we'd enter this with `pfd->events_requested == 0` in the first place.

Reply:
I guess it is an optimization to avoid making the `epoll_ctl` call, i.e., when we get an event we are looking for, we remove it from `events_requested` right away but don't do the ctl under the assumption that in many cases another event of that type may not occur again (e.g., we finish writing to a socket, we don't fill up the buffer on the last write we aren't expecting another EPOLLOUT event). Only if we _do_ get another event, which were weren't interested in, then we de-register interest in it?

Reply:
Basically I went down this train of thought to try to answer the question:

Why can't this change just be to extend the condition `if (events_to_remove)` to `if (events_to_remove || pfd->events_requested == 0)` since it seems like this would handle the error case too, but guess one reason not to do this is that would change the behavior above to eagerly derigister the fd instead of waiting until another event occurred.

So your change seems safer and smaller.

Reply:
Yes I went down the same train of thought and arrived at the same conclusion, i.e.: it potentially saves some syscalls. It relies on the fact that we serve all the ready futures and re-issue reads/writes again before polling the next time. If this wasn't epoll (and hence not immediately important) I think it would be interesting to confirm whether this actually happens "on average".

Ultimately this is were I hoped seastar folks could shed some light but yes as you say the change is smaller like this and doesn't remove that optimization.